go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -156,7 +161,9 @@ func (c *Controller) SyncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefine
 				ownedByUserDefinedNetwork(nad, udnGroupVersionKind, udn.UID) &&
 				controllerutil.ContainsFinalizer(nad, template.FinalizerUserDefinedNetwork) {
 
-				// TODO: verify no pod is using the NAD before removing finalizer from NAD
+				if err := c.verifyNetAttachDefNotInUse(nad); err != nil {
```

Comment:
we need some kind of infinite retry in case an error occurs here.
otherwise the UDN is deleted (deletionTimastamp is set), so it will not generate any more events, and then even when the pods using NAD get deleted, UDn won't be updated.
One way to solve this is to make a special error type for existing pods and then check this error type in `UserDefinedNetworkReconciler` and schedule next reconcile instead of returning an error, something like
```
	if errors.Is(syncErr, waitingForPodsError) {
		time.AfterFunc(1*time.Minute, func() {c.Controller.Reconcile(udn.Namespace+"/"+udn.Name)})
		return nil
	}
```

Reply:
Good catch! 
Thank you for the suggested snippet, done.

Reply:
testing this, it will create 2 timers because we will receive one more event for the same UDN on its update.
So we need smarter tracking, here is one idea:
add a new method to the existing controller.Reconciler interface
```
type Reconciler interface {
	Reconcile(key string)
	ReconcileAfter(key string, duration time.Duration)
...
}

func (c *controller[T]) ReconcileAfter(key string, duration time.Duration) {
	c.queue.AddAfter(key, duration)
}
```

and call this method instead of our own timer. Hopefully this will work better :)

Reply:
Done

From what I understand, since we use `workqueue.DefaultControllerRateLimiter` it would use an exponential rate-limiter, which start with very short backoff comparing to fixed 1 minute backoff, which makes the controller to act faster and the UDN & NAD dispose sooner.

From what I tested:
I created a UDN and a pod that use it.
- With AddAfter  1 minute delay, the UDN and NAD disposed after about 40 seconds since the pod was deleted.
- With AddRateLimited it took few seconds for the UDN and NAD to dispose.

WDYT?

Reply:
you are right that AddRateLimited works faster, but potentially removing the pods may take a very long time, and we don't want to spend this time retrying too many times. Also having a pre-defined period is nice to set users expectations of when it will be updated.

Reply:
the dangling object part and trying to notify users about expected behaviour will be there anyway.
If we use exponential backoff, it will retry with interval less than a minute for ~1.5 minutes after the UDN is deleted. If the pods are not deleted during that time, the exponential backoff will increase the retry interval and the behaviour will be similar to the 1m retry.
Our current `DefaultControllerRateLimiter` has the `maxDelay` of 1000 seconds, which is way too much to cleanup the UDN, so we will need to decrease it to at most 100 seconds just to match the expected behaviour for pod-related cleanup. At the same time, it means that if some UDN is failing to be created for some potentially valid reason, it will also be retried every 100 seconds instead of letting the retry interval grow to 1000 seconds. Also you won't need an extra method, it is enough to just return an error and the UDN will be put to the queue with rate limit, but you may need to change existing const `maxRetries` in controller package to make it configurable per-controller and just set it to 0 in this case as we want infinite retry.
I think it is not too bad, but is definitely worth a comment if you want to go that way.

Reply:
Thanks for explaining, I am OK with the current state.
I think we can optimize it in a follow up PR, if needed, make sense?

Reply:
sure, I am fine with either way.